shared/BPCSharedComponent/DSound.cs

```diff
@@ -313,47 +320,62 @@ public static void setRootDirectory(String root)
+		/// The code has finally been improved thanks to the MonoGame framework code: https://github.com/MonoGame/MonoGame
 		/// </summary>
 		/// <param name="sound">The sound to pan.</param>
 		/// <param name="pan">The value by which to pan the sound. -1.0f is completely left, and 1.0f is completely right. 0.0f is center.</param>
 		public static void setPan(ExtendedAudioBuffer sound, float pan)
 		{
-			SpeakerConfiguration mask = (SpeakerConfiguration)mainMasteringVoice.ChannelMask;
-			float[] outputMatrix = new float[8];
-			float left = 0.5f - pan / 2;
-			float right = 0.5f + pan / 2;
-			switch(mask) {
-				case SpeakerConfiguration.mono:
-					outputMatrix[0] = 1.0f;
-					break;
-				case SpeakerConfiguration.stereo:
-				case SpeakerConfiguration.twoPointOne:
-				case SpeakerConfiguration.surround:
-					outputMatrix[0] = left;
-					outputMatrix[1] = right;
-					break;
-				case SpeakerConfiguration.quad:
-					outputMatrix[0] = outputMatrix[2] = left;
-					outputMatrix[1] = outputMatrix[3] = right;
-					break;
-				case SpeakerConfiguration.fourPointOne:
-					outputMatrix[0] = outputMatrix[3] = left;
-					outputMatrix[1] = outputMatrix[4] = right;
-					break;
-				case SpeakerConfiguration.fivePointOne:
-				case SpeakerConfiguration.sevenPointOne:
-				case SpeakerConfiguration.fivePointOneSurround:
-					outputMatrix[0] = outputMatrix[4] = left;
-					outputMatrix[1] = outputMatrix[5] = right;
-					break;
-				case SpeakerConfiguration.sevenPointOneSurround:
-					outputMatrix[0] = outputMatrix[4] = outputMatrix[6] = left;
-					outputMatrix[1] = outputMatrix[5] = outputMatrix[7] = right;
-					break;
+			SpeakerConfiguration mask;
+			if (mainSoundDevice.Version == XAudio2Version.Version27)
+			{
+				WaveFormatExtensible deviceFormat = mainSoundDevice.GetDeviceDetails(0).OutputFormat;
+				mask = (SpeakerConfiguration)deviceFormat.ChannelMask;
 			}
+			else
+				mask = (SpeakerConfiguration)mainMasteringVoice.ChannelMask;
 			VoiceDetails soundDetails = sound.getVoiceDetails();
 			VoiceDetails masteringDetails = mainMasteringVoice.VoiceDetails;
+			int srcChannelCount = soundDetails.InputChannelCount;
+			int dstChannelCount = masteringDetails.InputChannelCount;
+			// Create an array to hold the output matrix. Warning : the minimum size of the output matrix is the number of channels in the source voice times the number of channels in the output voice.
+			// Note that the outputMatrix indices are placed in the same order as the SharpDX.Multimedia.Speakers enum.  
+			// Don't forget there are two times more cells in the matrix if the source sound is stereo)
+			float[] outputMatrix = new float[srcChannelCount * dstChannelCount];
+			Array.Clear(outputMatrix, 0, outputMatrix.Length);
+			// From there, we'll hope that the sound file is either mono or stereo. If the WAV had more than 2 channels, it would be to difficult to handle. 
```

Comment:
Minor type-o here, please change to "too" and not to".

Reply:
Pushed two new commits ... sorry for that. I made a typo when fixing the initial typo ... lol